### PR TITLE
fix: add RFC 1035 length limits to domain regex

### LINKED ISSUE #5965: regexes.domain is missing RFC 1035 length limits
## Context

`regexes.domain` (used by `z.httpUrl()` for the URL host check) currently has no length anchors:

```ts
export const domain: RegExp = /^([a-zA-Z0-9](?:[a-zA-Z0-9_-]{0,61}[a-zA-Z0-9])?\.)+[a-zA-Z]{2,}$/;
```

vs. `regexes.hostname` (used by `z.hostname()`), which does enforce the RFC 1035 limits:

```ts
export const hostname: RegExp =
  /^(?=.{1,253}\.?$)[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?(?:\.[a-zA-Z0-9](?:[-0-9a-zA-Z]{0,61}[0-9a-zA-Z])?)*\.?$/;
```

So `z.httpUrl()` accepts hosts that exceed the 253-character total or have non-TLD labels longer than 63 characters, while `z.hostname()` rejects them.

## Question

Is this an intentional looseness for `z.httpUrl()` (since RFC 3986 / WHATWG URL doesn't itself impose DNS length limits, even though the underlying DNS does), or a gap worth closing?

If the latter, the fix is small — adding a `(?=.{1,253}\.?$)` lookahead and capping the TLD with a `{2,63}` quantifier. Happy to follow up with a PR if there's interest.

## Note

This is independent from #5963 (underscore support, currently open), but it'd be cleaner to land that first to avoid noisy rebases.

### PR DESCRIPTION
**Pull request by mixelburg**
## Problem

The `regexes.domain` regex (used internally by `z.httpUrl()` for hostname validation) was missing RFC 1035 length limits that the `regexes.hostname` regex already enforces:

- No 253-character total length limit on the domain
- No upper bound on TLD label length (`{2,}` vs `{2,63}`)

This meant `z.httpUrl()` accepted domains exceeding the 253-character DNS limit and TLDs longer than 63 characters, while `z.hostname()` correctly rejected them.

## Fix

Added RFC 1035 length constraints to `regexes.domain`:

- `(?=.{1,253}\.?$)` lookahead — enforces the 253-character total domain length limit
- `{2,63}` on the TLD label — caps the last label at 63 characters per RFC 1035

## Testing

- All existing 3811 tests pass across 339 test files
- Added 2 new test cases: TLD >63 chars rejected, domain >253 chars rejected

Fixes #5965

### CODE REVIEW

**Comment by mixelburg** [comment]
Hi! Is there anything else needed from my side on this PR?

**Comment by mixelburg** [comment]
Updated the two stale regex references:\n- `packages/docs/content/api.mdx:382` — sync'd embedded regex literal with the current `z.regexes.domain` pattern\n- `packages/zod/src/v4/classic/tests/string.test.ts:433` — updated comment regex to match the new RFC 1035 limits

File: packages/zod/src/v4/core/regexes.ts [hunk 1]
```diff
@@ -84,7 +84,7 @@ export const base64url: RegExp = /^[A-Za-z0-9_-]*$/;
 export const hostname: RegExp =
   /^(?=.{1,253}\.?$)[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?(?:\.[a-zA-Z0-9](?:[-0-9a-zA-Z]{0,61}[0-9a-zA-Z])?)*\.?$/;
 
-export const domain: RegExp = /^([a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?\.)+[a-zA-Z]{2,}$/;
+export const domain: RegExp = /^(?=.{1,253}\.?$)([a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?\.)+[a-zA-Z]{2,63}$/;
```
[the remark this answers is not in the document]
Comment (mixelburg) [reply packages/zod/src/v4/core/regexes.ts:88]: Good suggestion — added the RFC 1035 annotation in 3df7fb6.

**Review by wahajahmed010 — COMMENTED** [review]
## Review

Nice catch — the inconsistency between `regexes.hostname` (already RFC 1035-compliant) and `regexes.domain` was a latent footgun in `z.httpUrl()`'s default. Aligning them is the right move.

Substantive notes:

- `(?=.{1,253}\.?$)` is correct for the total-length constraint. `.` matches everything except newline by default in JS (no `s` flag needed here, no `m` flag set), so embedded newlines correctly fail the lookahead and the whole regex. Good.
- `{2,63}` on the TLD matches RFC 1035 §2.3.4 (label length 1–63 octets). With the leading label constraint `{0,61}` from the rest of the regex, all labels are also capped — consistent with `hostname`.
- **Behavioural change worth flagging**: any caller passing a string longer than 253 chars or with a >63-char label that was previously accepted by `z.string().regex(z.regexes.domain)` will now fail validation. That's the correct behaviour, but it's still a change — worth a CHANGELOG entry under `zod/v4`. If the project does a minor bump soon, an `[next]` note is appropriate.
- Test additions cover the two failure modes (TLD >63 chars, total >253 chars). Suggest two more boundary cases to lock the contract:
  - exactly 63-char TLD → should pass (`{2,63}` inclusive upper bound)
  - exactly 253-char total length → should pass (`{1,253}` inclusive)
  
  These guard against off-by-one if anyone ever changes the quantifiers.
- One small nit on the diff: the JSDoc comment `/** RFC 1035: labels ≤63 chars, total ≤253 chars, TLD 2–63 chars. */` slightly undersells what the regex enforces — the leading label is also bounded to `[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?` (≤63 chars), same as `hostname`. Consider matching that wording for consistency, or just point readers at `regexes.hostname`'s comment.

Otherwise solid. Approving once the boundary-case tests land or in a follow-up.

**Comment by mixelburg** [comment]
Closing due to inactivity — 3 months with no human review or engagement. Happy to reopen if there's interest.
